### PR TITLE
New terrain support - part 1

### PR DESCRIPTION
**Pull request by Nordsoft91**
As change is huge, I decided to pass it through multiple reviews to avoid potential merge conflicts.

Current part supports new terrains and new castles.

This is major change, which means that backward compatibility is violated at some places and code may not be compatible with some mods. However, all features shall work as expected.

Please, use attached mod to test new terrain.

[test_map_terrains-mod.zip](https://github.com/vcmi/vcmi/files/8881179/test_map_terrains-mod.zip)

### CODE REVIEW

File: .github/workflows/github.yml
Comment (Nordsoft91) [inline .github/workflows/github.yml:133]: @nullkiller , if we go with same approach (feature branch), we can keep it. Otherwise will be reverted

File: AI/Nullkiller/Pathfinding/Actors.cpp
Comment (Nordsoft91) [inline AI/Nullkiller/Pathfinding/Actors.cpp:73]: std::exception doesn't have non-default constructor. It's bug in CL implementation

File: client/mapHandler.cpp
Comment (Nordsoft91) [inline client/mapHandler.cpp:149]: will be removed

File: lib/mapping/CMapDefines.h
Comment (alexvins) [inline lib/mapping/CMapDefines.h:83]: This line completely breaks backward compatibility for saves. If that is intentional all compatibility handling code should be cleaned up, `SERIALIZATION_VERSION` & `MINIMAL_SERIALIZATION_VERSION` constants should be changed.
(`SERIALIZATION_VERSION` should be changed even if compatibility is retained)

**Review by alexvins — CHANGES_REQUESTED** [review]

File: .github/workflows/github.yml
Comment (nullkiller) [reply .github/workflows/github.yml:133]: A new branch was created and the PR was retargeted

File: lib/mapping/CMapDefines.h
Comment (nullkiller) [reply lib/mapping/CMapDefines.h:83]: I think for such a big feature it is ok to completely break backward compatibility. Any objections? We will merge it to develop only after it is completely ready. For now feature builds will be ok.

Comment (ShubusCorporation) [reply lib/mapping/CMapDefines.h:83]: I think it's not hard to ensure backward compatibility at this point: if version of BS < current => read old ETerrainType and create new CTerrainType instance using that old -style value, otherwise, just read  CTerrainType object

So, please think about it before merge...

**Review by nullkiller — CHANGES_REQUESTED** [review]

File: lib/mapping/MapFormatJson.cpp
Comment (Nordsoft91) [reply lib/mapping/MapFormatJson.cpp:963]: How can I test it?

File: lib/mapping/CMapDefines.h
Comment (Nordsoft91) [reply lib/mapping/CMapDefines.h:83]: @alexvins , thank you, didn't know about it.
@nullkiller , I think for feature branch we can proceed without backward compatibility, but before merging into develop I'll try to implement something similar to what @ShubusCorporation suggested

File: client/mapHandler.cpp
Comment (Nordsoft91) [reply client/mapHandler.cpp:1447]: You are right, I tried to get rid from id() method, but it's tricky

File: lib/CTerrainType.h
Comment (Nordsoft91) [reply lib/CTerrainType.h:22]: It works: rmg doesn't use new terrains while mod is disabled, but if you try to load the game with asphalt it crashes:(

File: lib/CTownHandler.h
Comment (Nordsoft91) [reply lib/CTownHandler.h:389]: CTownHandler.cpp, 29-31

File: lib/CTerrainType.cpp
Comment (Nordsoft91) [reply lib/CTerrainType.cpp:120]: id is temporary method, I'm thinking on how to get rid of it. Initially I used it almost everywhere, now it's only few places.

File: lib/mapping/MapFormatJson.cpp
Comment (alexvins) [inline lib/mapping/MapFormatJson.cpp:963]: - `typeCode` is defined in  `namespace TerrainDetail` in this file. It should be moved to terrains.json config and handled somehow.
- `src` is not even a terrain name, it 2 letter terrain code + "terrain view" + "terrain flip"

File: lib/mapping/MapFormatJson.cpp
Comment (alexvins) [inline lib/mapping/MapFormatJson.cpp:1296]: Terrain code is not the first 2 letters of terrain name. See `TerrainDetail::terrainCodes`

**Review by alexvins — CHANGES_REQUESTED** [review]

File: lib/mapping/MapFormatJson.cpp
Comment (Nordsoft91) [reply lib/mapping/MapFormatJson.cpp:1296]: @alexvins , I have no way to debug this code. Please, help with obtaining maps in json format

Comment (alexvins) [reply lib/mapping/MapFormatJson.cpp:1296]: Have you run the test suite?

Comment (nullkiller) [reply lib/mapping/MapFormatJson.cpp:1296]: @alexvins in order to support new terrains everywhere we have to patch the redactor as well. Could you help with it in a later phase when we have json format stable? Or should I consider supporting this activity?

Comment (nullkiller) [reply lib/mapping/MapFormatJson.cpp:1296]: Our redactor is only the way to create a new map with new terrains. Otherwise we can relay only on autogenerated maps

Comment (alexvins) [reply lib/mapping/MapFormatJson.cpp:1296]: > Our redactor is only the way to create a new map with new terrains. Otherwise we can relay only on autogenerated maps

I mean test suite covers map generation, saving to json and loading back It should be enough to test/debug this code

Comment (nullkiller) [reply lib/mapping/MapFormatJson.cpp:1296]: Yes but this does not solve the problem that there is no tool to create an authored map with new terrains. Also later new towns might start using new terrains in their configs and so on so the redactor patch is inevitable at the end.

Comment (nullkiller) [reply lib/mapping/MapFormatJson.cpp:1296]: Probably it is too early to consider this activity so. There is a lot of other work and now we can just use the test suit

Comment (nullkiller) [reply lib/mapping/MapFormatJson.cpp:1296]: I will try to build vcmi_editor from source so and consider to publish a new release binaries for windows

Comment (Nordsoft91) [reply lib/mapping/MapFormatJson.cpp]: @alexvins , do you mean unit tests? If so, unfortunately they are broken even on develop branch.
https://h3vcmi.slack.com/archives/C2J9QFUDS/p1652881185132249

Comment (Nordsoft91) [reply lib/mapping/MapFormatJson.cpp]: What is sense to keep functionality which is not used and not supported? If it used, some maps should be available somewhere:)

Comment (nullkiller) [reply lib/mapping/MapFormatJson.cpp]: Might be that modders are not interested in creating custom authored maps for VCMI only?

Comment (Nordsoft91) [reply lib/mapping/MapFormatJson.cpp]: Right, probably, there should be two-way converter

Comment (nullkiller) [reply lib/mapping/MapFormatJson.cpp]: Two-way not possible. H3M can not handle our new objects. One way exists, I've sent you a few maps to slack

File: lib/mapping/MapFormatJson.cpp
Comment (Nordsoft91) [reply lib/mapping/MapFormatJson.cpp]: fixed

File: lib/mapObjects/CObjectClassesHandler.cpp
Comment (Nordsoft91) [reply lib/mapObjects/CObjectClassesHandler.cpp:198]: It was needed to allow adding templates for existing objects, e.g. mines.
I'm not sure how it works exactly, need testing

File: lib/mapping/CMapDefines.h
Comment (Nordsoft91) [reply lib/mapping/CMapDefines.h]: Investigation shown that probably we should sacrifice backward compatibility for saves

Comment (nullkiller) [reply lib/mapping/CMapDefines.h]: Yes it is highly likely we should. Anyway we can merge this PR anytime you want as it does not target develop. Just ensure you track somehow all important points spotted that will be addressed later so that they are not get lost.

File: lib/JsonNode.cpp
Comment (Nordsoft91) [reply lib/JsonNode.cpp:724]: I put a comment earlier, see line 723

File: lib/mapping/MapFormatJson.cpp
Comment (val-gaav) [reply lib/mapping/MapFormatJson.cpp]: Custom map forces users to use certain set mods, while people like to select the mods they use themselves. This is probably why many people may prefer random maps with their own rules/mods. Anyway editor and map format is of course needed and is a must have, it is just that the only way for it to be in real use is if VCMI gets some big mod.   Probably people from MDT team had some plans to do custom maps/campaigns, though it was never published or continued as far as I know.

File: lib/JsonNode.cpp
Comment (nullkiller) [reply lib/JsonNode.cpp:724]: I am afraid it is quite rare case which might be even never used. Some creature bonus which work on certain terrain only. Usually it is native terrain and exact terrain name will not be specified.

Comment (Alucard648) [reply lib/JsonNode.cpp:724]: OH3 Nomads in army allows hero to ignore terrain movement penalty on desert.

**Pull request merged**
